jason/src/peer/repo.rs

```diff
@@ -76,14 +76,10 @@ impl Repository {
             loop {
                 delay_for(Duration::from_secs(1).into()).await;
 
+                let peers =
+                    peers.borrow().values().cloned().collect::<Vec<_>>();
                 future::join_all(
-                    peers
-                        .borrow()
-                        .values()
-                        .cloned()
-                        .collect::<Vec<_>>()
-                        .iter()
-                        .map(|peer| peer.scrape_and_send_peer_stats()),
```

Comment:
@tyranron ,

Вот, кстати, место, где ссылка протекала в футуру. Достаточно контринтуитивно.

Reply:
@alexlapa yup, there are quite problems with `RefCell`/`Mutex` leaking into `Future`s. I had numerous problems with that.

The one rule that helps me a lot and quite good eliminates such footguns is "__always__ keep guards alive as least as possble". That's why I usually wrap guard usage with explicit block like:
```rust
let peers = {
    peers.borrow().values().cloned().collect::<Vec<_>>()
};
future::join_all(peers.iter() .... )
```
Or even often refactor it to a separate getter like `fn cloned_peers(&self) -> Vec<_>`.

Such preventive programming way works quite good so far, and I haven't any footguns recently with it.

I wish Rust could do something with it, but can't even imagine how it could have been solved to disjoint such situations at all.